### PR TITLE
WIP An initial implementation of cells

### PR DESCRIPTION
**Pull request by jasongrout**
This relies on https://github.com/jupyter/jupyter-js-output-area/pull/24

### CODE REVIEW

File: src/CellViewModel.ts [hunk 1]
```diff
@@ -0,0 +1,348 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+import {
```
Comment (blink1073) [inline src/CellViewModel.ts:9]: Missing blank line.

File: src/CellViewModel.ts [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  /**
+   * Whether the cell is marked for applying commands
+   */
+  //marked: boolean;
+
+  /**
+   * Run the cell.
+   * 
+   * This basically means: do the right thing with the input.
+   * Perhaps this should be in a subclass ExecutableCell.  Do we want to have a run() method on raw cells?
```
Comment (blink1073) [inline src/CellViewModel.ts:119]: I'm fine with it being a no-op for raw cells.

File: src/CellViewModel.ts [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+  static stateChangedSignal = new Signal<IBaseCellViewModel, IChangedArgs<any>>();
+
+
+  /**
+   * A signal emitted when the state of the model changes.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[stateChangedSignal]].
+   */
+  get stateChanged(): ISignal<IBaseCellViewModel, IChangedArgs<any>> {
```
Comment (blink1073) [inline src/CellViewModel.ts:192]: Put all of the static stuff first.

File: src/CellViewModel.ts [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+   *
+   * #### Notes
+   * This is a pure delegate to the [[inputProperty]].
+   */
+  set input(value: IInputAreaViewModel) {
+    BaseCellViewModel.inputProperty.set(this, value);
+  }
+
+  /**
+   * Run the cell
```
Comment (blink1073) [inline src/CellViewModel.ts:227]: Might want to add a note about the default implementation being a no-op.

File: src/CellViewModel.ts [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+   */
+  run(): void {
+  }
+
+  /**
+   * The type of cell.
+   */
+  type: CellType;
```
Comment (blink1073) [inline src/CellViewModel.ts:235]: Why not a read-only getter?

File: src/CellViewModel.ts [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+
+/**
+ * An implementation of a code cell viewmodel.
+ */
+export
+class CodeCellViewModel extends BaseCellViewModel implements ICodeCellViewModel {
+
+  /**
+   * A signal emitted when the state of the model changes.
```
Comment (blink1073) [inline src/CellViewModel.ts:246]: Wrong docstring.

File: src/CellViewModel.ts [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+   *
+   * **See also:** [[stateChanged]]
+   */
+  static executeRequestSignal = new Signal<ICodeCellViewModel, string>();
+
+
+  /**
+   * A signal emitted when the cell is requesting execution.
+   * 
+   * TODO: Do we need this execute signal?
```
Comment (blink1073) [inline src/CellViewModel.ts:256]: I believe we do.

File: src/CellViewModel.ts [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+   *
+   * #### Notes
+   * This is a pure delegate to the [[stateChangedSignal]].
```
Comment (blink1073) [inline src/CellViewModel.ts:259]: Stale.

File: src/CellViewModel.ts [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+   */
+  get executeRequest() {
+    return CodeCellViewModel.executeRequestSignal.bind(this);
+  }
+
+  /**
+  * A property descriptor holding the output area view model.
+  * 
+  * TODO: Do we need this execute signal?
```
Comment (blink1073) [inline src/CellViewModel.ts:268]: Remnant.

File: src/CellViewModel.ts [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+  * **See also:** [[output]]
+  */
+  static outputProperty = new Property<CodeCellViewModel, IOutputAreaViewModel>({
+      name: 'output',
```
Comment (blink1073) [inline src/CellViewModel.ts:272]: 4 spaces?

File: src/CellViewModel.ts [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+
+
+  /**
+   * Get the output area view model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[outputProperty]].
+   */
+  get output() { 
+      return CodeCellViewModel.outputProperty.get(this);
```
Comment (blink1073) [inline src/CellViewModel.ts:284]: idem

File: src/CellViewModel.ts [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+   *
+   * #### Notes
+   * This is a pure delegate to the [[outputProperty]].
+   */
+  set output(value: IOutputAreaViewModel) {
+      CodeCellViewModel.outputProperty.set(this, value);
+  }
+
+  /**
+   * Run the cell
```
Comment (blink1073) [inline src/CellViewModel.ts:298]: Missing a period.

File: src/CellWidget.ts [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+import {
+  Widget, Panel
+} from 'phosphor-widget';
+
+import {
+    ICodeCellViewModel, IMarkdownCellViewModel
+} from './CellViewModel';
+
+import {
```
Comment (blink1073) [inline src/CellWidget.ts:17]: These three imports go before the phosphor ones (alphabetically).

File: src/CellWidget.ts [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    InputAreaWidget, IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+    OutputAreaWidget, IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import * as marked from 'marked';
+
+/**
```
Comment (blink1073) [inline src/CellWidget.ts:27]: 2 lines.

File: src/CellWidget.ts [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+  constructor(model: ICodeCellViewModel) {
+    super();
+    this.addClass('jp-Cell');
+    this.addClass('jp-CodeCell');
+    this._model = model;
+    this.input = new InputAreaWidget(model.input);
+    this.output = new OutputAreaWidget(model.output);
+    this.children.assign([this.input, this.output]);
+
+    this.output = new OutputAreaWidget(model.output);
```
Comment (blink1073) [inline src/CellWidget.ts:45]: Repeated lines.

File: src/CellWidget.ts [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    this.children.assign([this.input, this.output]);
+    model.stateChanged.connect(this.modelStateChanged, this);
+  }
+
+  /**
+   * Update the input area, creating a new input area
+   * widget and detaching the old one.
+   */
+  updateInputArea(input: IInputAreaViewModel) {
```
Comment (blink1073) [inline src/CellWidget.ts:54]: Should these be protected?

File: src/CellWidget.ts [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+   */
+  editInput() {
+    this.children.set(0, this.input);
+  }
+
+  /**
+   * Update the input area, creating a new input area
+   * widget and detaching the old one.
+   */
+  updateInputArea(input: IInputAreaViewModel) {
```
Comment (blink1073) [inline src/CellWidget.ts:143]: Protected?

**Comment by blink1073** [comment]
Ran the example, looks good, I'd recommend adding this:

``` diff
diff --git a/package.json b/package.json
index 0c28fd5..c264851 100644
--- a/package.json
+++ b/package.json
@@ -58,6 +58,11 @@
     "lib/*.js",
     "lib/*.d.ts"
   ],
+  "system": {
+    "npmIgnore": [
+      "devDependencies"
+    ]
+  },
   "author": "Project Jupyter",
   "license": "BSD-3-Clause",
   "bugs": {
```

**Comment by jasongrout** [comment]
@blink1073 - Thanks for looking at this.  I forgot to label this WIP.  It's definitely not ready for scrutiny about documentation, whitespace, etc.  I'm moving on to the notebook repo, which will test the design of this.  When the basic system works, I'm coming back to go over the documentation, whitespace, etc.

**Comment by blink1073** [comment]
Ha, fair enough.

**Comment by jasongrout** [comment]
Superseded by #4

**Pull request closed without merging**